src/register/mcause.rs

```diff
@@ -1,34 +1,75 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
 //! mcause register
 
-/// mcause register
-#[derive(Clone, Copy, Debug)]
-pub struct Mcause {
-    bits: usize,
+ro_csr!(mcause, usize);
+
+register_bitfields![usize,
+    #[cfg(target_pointer_width = "32")]
+    pub mcause [
+        cause OFFSET(0) NUMBITS(31) [],
+        is_interrupt OFFSET(31) NUMBITS(1) [],
+    ],
+
+    #[cfg(target_pointer_width = "64")]
+    pub mcause [
+        cause OFFSET(0) NUMBITS(63) [],
+        is_interrupt OFFSET(63) NUMBITS(1) [],
+    ],
+];
+
+/// Get the cause of the latest interrupt
+#[inline]
+pub fn cause() -> Trap {
+    let cause: usize = read_field(mcause::cause);
+    if is_interrupt() {
+        Trap::Interrupt(Interrupt::from(cause))
+    } else {
+        Trap::Exception(Exception::from(cause))
+    }
+}
+
+/// Returns `true` if the last trap was caused by an interrupt, otherwise
+/// `false`.
+#[inline]
+pub fn is_interrupt() -> bool {
+    is_set(mcause::is_interrupt)
 }
 
-/// Trap Cause
+/// Returns `true` if the last trap was caused by an exception, otherwise
+/// `false`.
+#[inline]
+pub fn is_exception() -> bool {
+    !is_interrupt()
+}
+
+/// Enum wrapping both potential trap causes
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 pub enum Trap {
     Interrupt(Interrupt),
     Exception(Exception),
 }
 
-/// Interrupt
+// TODO: Maybe add a feature that changes these enums to use the actual values,
+// and the conversion functions to use core::mem::transmute(). Assuming that
+// the target platform doesn't use the reserved or custom interrupt lines this
+// should be safe. This would only be worthwhile if the compiler doesn't already
+// optimize the conversions out.
+/// Interrupt causes
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 pub enum Interrupt {
-    UserSoft,
```

Comment:
Why have the user interrupt types been removed?

Reply:
It doesn't look like the privilege spec actually supports these, unless there was another extension that I am unaware of that adds them. I've been looking at [version 20211203](https://github.com/riscv/riscv-isa-manual/releases/download/Priv-v1.12/riscv-privileged-20211203.pdf).

Reply:
So they were in older versions of the spec, but got removed.  I would actually be in favor of keeping them in the crate though.

Reply:
It looks like there was an `N` extension that added these, but it has been removed since version 1.11. Not sure if it's worth keeping this or not then.